File: server/call_mailbox.py
"""Durable, idempotent outgoing-call requests."""
import re
import time

TTL_MS = 120000
FINAL = {"failed", "ended", "expired"}


def now_ms():
    return int(time.time() * 1000)


def valid(command, now=None):
    now = now_ms() if now is None else now
    created = command.get("created_at") if isinstance(command, dict) else None
    return (isinstance(command, dict)
            and isinstance(command.get("id"), str) and re.fullmatch(r"[a-f0-9-]{36}", command["id"])
            and isinstance(command.get("to"), str) and re.fullmatch(r"\+?[0-9]{3,20}", command["to"])
            and (command.get("subscription_id") is None or
                 type(command["subscription_id"]) is int and 0 <= command["subscription_id"] <= 2147483647)
            and type(created) is int and now - TTL_MS < created <= now + 300000)


def submit(records, command, now=None):
    now = now_ms() if now is None else now
    if not isinstance(command, dict):
        return 400, None
    old = next((x for x in records if x["id"] == command.get("id")), None)
    keys = ("to", "subscription_id", "created_at")
    if old:
        return (200, old) if all(old.get(k) == command.get(k) for k in keys) else (409, None)
    if not valid(command, now):
        return 400, None
    if any(x["status"] not in FINAL for x in records):
        return 409, None
    record = {k: command.get(k) for k in ("id", "to", "subscription_id", "created_at")}
    record.update(status="queued", detail="", expires_at=command["created_at"] + TTL_MS)
    records[:] = [x for x in records if x.get("created_at", 0) > now - 172800000]
    records.append(record)
    return 200, record


def expire(records, now=None):
    now = now_ms() if now is None else now
    for record in records:
        if record["status"] == "queued" and record["expires_at"] <= now:
            record.update(status="expired", detail="Sender did not place the call before it expired")


def update(records, result):
    allowed = {"dialing", "connected", "failed", "ended"}
    if not isinstance(result, dict) or result.get("status") not in allowed:
        return False
    record = next((x for x in records if x["id"] == result.get("id")), None)
    if record is None:
        return False
    if record["status"] in FINAL:
        return True
    record.update(status=result["status"], detail=str(result.get("detail", ""))[:240])
    return True


def cancel_active(records, detail="Receiver started a new call"):
    changed = False
    for record in records:
        if record["status"] not in FINAL:
            record.update(status="ended", detail=detail[:240])
            changed = True
    return changed

File: server-workers/src/call-mailbox.mjs
export const CALL_TTL_MS = 120000;
const FINAL = new Set(['failed', 'ended', 'expired']);
export function validCall(c, now = Date.now()) {
  return c && typeof c.id === 'string' && /^[a-f0-9-]{36}$/.test(c.id) &&
    typeof c.to === 'string' && /^\+?[0-9]{3,20}$/.test(c.to) &&
    (c.subscription_id == null || Number.isInteger(c.subscription_id) && c.subscription_id >= 0 && c.subscription_id <= 2147483647) &&
    Number.isSafeInteger(c.created_at) && c.created_at <= now + 300000 && c.created_at > now - CALL_TTL_MS;
}
export function submitCall(records, command, now = Date.now()) {
  const old = records.find(x => x.id === command?.id);
  if (old) return old.to === command.to && (old.subscription_id ?? null) === (command.subscription_id ?? null) &&
    old.created_at === command.created_at ? {status: 200, records, command: old} : {status: 409};
  if (!validCall(command, now)) return {status: 400};
  if (records.some(x => !FINAL.has(x.status))) return {status: 409};
  const retained = records.filter(x => x.created_at > now - 172800000);
  const next = {id: command.id, to: command.to, subscription_id: command.subscription_id ?? null,
    created_at: command.created_at, expires_at: command.created_at + CALL_TTL_MS, status: 'queued', detail: ''};
  return {status: 200, records: [...retained, next], command: next};
}
export function expireCalls(records, now = Date.now()) {
  return records.map(x => x.status === 'queued' && x.expires_at <= now
    ? {...x, status: 'expired', detail: 'Sender did not place the call before it expired'} : x);
}
export function updateCall(records, result) {
  if (!result || !['dialing', 'connected', 'failed', 'ended'].includes(result.status)) return false;
  const record = records.find(x => x.id === result.id);
  if (!record) return false;
  if (FINAL.has(record.status)) return true;
  record.status = result.status;
  record.detail = typeof result.detail === 'string' ? result.detail.slice(0, 240) : '';
  return true;
}
